docs/architecture/psa-migration/strategy.md

```diff
@@ -0,0 +1,385 @@
+useful uses of `MBEDTLS_PSA_CRYPTO_CONFIG`. The point of
+`MBEDTLS_PSA_CRYPTO_CONFIG` is to be able to build a library with support for
+an algorithm through a PSA driver only, without building the software
+implementation of that algorithm. But then the TLS code would consider the
+algorithm unavailable.
+
+This is tracked in https://github.com/ARMmbed/mbedtls/issues/3674 and
+https://github.com/ARMmbed/mbedtls/issues/3677. But now that I look at it with
+fresh eyes, I don't think the approach we were planning to use would actually
+works. This needs more design effort.
+
+This is something we need to support eventually, and several partners want it.
+I don't know what the priority is for `MBEDTLS_USE_PSA_CRYPTO` between
+improving driver support and covering more of the protocol. It seems to me
+that it'll be less work overall to first implement a good architecture for
+`MBEDTLS_USE_PSA_CRYPTO + MBEDTLS_PSA_CRYPTO_CONFIG` and then extend to more
+protocol featues, because implementing that architecture will require changes
+to the existing code and the less code there is at this point the better,
+whereas extending to more procotol features will require the same amount of
+work either way.
+
+### `MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER`
+
+When `MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER` is enabled, the library is
+built for use with an RPC server that dispatches PSA crypto function calls
+from multiple clients. In such a build, all the `psa_xxx` functions that take
+would normally take a `psa_key_id_t` as argument instead take a structure
+containing both the key id and the client id. And so if e.g. a TLS function
+calls `psa_import_key`, it would have to pass this structure, not just the
+`psa_key_id_t` key id.
+
+A solution is to use `mbedtls_svc_key_id_t` throughout instead of
+`psa_key_id_t`, and use similar abstractions to define values.
+
+That's what we're implemementing in early 2022, see
+https://github.com/ARMmbed/mbedtls/issues/5259
+
+### Backwars compatibility issues with making it always on
+
+1. Existing applications may not be calling `psa_crypto_init()` before using
+   TLS, X.509 or PK. We can try to work around that by calling (the relevant
+part of) it ourselves under the hood as needed, but that would likely require
+splitting init between the parts that can fail and the parts that can't (see
+https://github.com/ARM-software/psa-crypto-api/pull/536 for that).
+2. It's currently not possible to enable `MBEDTLS_PSA_CRYPTO_C` in
+   configurations that don't have `MBEDTLS_ENTROPY_C`, and we can't just
+auto-enable the latter, as it won't build or work out of the box on all
+platforms. There are two kinds of things we'd need to do if we want to work
+around that:
+   1. Make it possible to enable the parts of PSA Crypto that don't require an
+      RNG (typically, public key operations, symmetric crypto, some key
+management functions (destroy etc)) in configurations that don't have
```

Comment:
That's on the planned feature list anyway. For example, to build a bootloader that needs to verify a signature and doesn't have a TRNG driver.

But as far as `MBEDTLS_USE_PSA_CRYPTO` is concerned, this is only relevant for X.509 verification, not for the rest, especially TLS.

Reply:
I've been thinking about it, and a large part of the crypto operations in TLS (in lines of code, not execution time) is symmetric crypto: hash/mac, cipher/aead. Most of the functions in these areas shouldn't require an RNG, except a few specific ones (such as `psa_aead_generate_nonce()`) that are not used by TLS and that (I think) should be easy to exclude from RNG-less builds.

I think there would already be some value in making `MBEDTLS_USE_PSA_CRYPTO` non-optional for all the symmetric crypto in X.509 and TLS, for an amount of effort that's probably much lower than for asymmetric crypto. So, that could probably be a very nice intermediate goal.

Reply:
I'm reluctant to allow any crypto that uses a secret key without an RNG, even if it's currently implemented without randomization. It means promising we'll never have any countermeasures involving randomization. It also means putting a separation between `generate_{iv,nonce}` and other cipher/AEAD functions, which makes the API less easy to understand.